core/app/models/spree/promotion_code_builder.rb

```
@@ -0,0 +1,51 @@
+class Spree::PromotionCodeBuilder
+  attr_reader :promotion
+
+  class_attribute :default_random_code_length
+  self.default_random_code_length = 6
+
+  def initialize(promotion_attrs:, base_code:, number_of_codes:, user: nil)
+    @promotion = Spree::Promotion.new(promotion_attrs)
+    @base_code = base_code
+    @number_of_codes = number_of_codes
+    @user = user
+  end
+
+  def perform
+    build_promotion_codes if @base_code && @number_of_codes
```

There isn't much here in the way of helping the programmer use this correctly.

If neither number of codes nor base code are passed in, should this builder even be used? If neither are passed in, do we want to do nothing and save the promotion?

I think it makes more sense to inherit ActiveModel::Model and validate presense of base code and number of codes, and error our rather than silently doing nothing, considering the purpose of this class.